---
title: "CRISPR L1s differential expression analysis of TEs"
output: html_notebook
---

## Here we perform differential expression analysis (DEA) of genes for our CRISPR experiments targeting young L1 elements.

The experimental design is CRISPR inhibition on iPSC where L1s are usually highly expressed, and CRISPR activation on neural progenitor cells.
The two experiments were sequenced in two sequencing runs each (three sequencing runs total). 

Before we start, I load libraries and define some helper functions to visualize the effect as a mean plot (credits to Per Brattås! who wrote the original code for some of these)

```{r class.source = 'fold-hide'}
## getSignName ##
# Get significantly different gene names. 
# Taken from source code of the package deseqAbstraction which is no longer available on github.
# Credits to Per L. Brattås
# Parameters:
# x = results object from deseq
# p = padj threshold for significance
# l = log2FC threshold for significance
getSignName <- function(x,p,l=0) {
  up <- x[!is.na(x$padj) & x$padj < p & x$log2FoldChange > l,]
  down <- x[!is.na(x$padj) & x$padj < p & x$log2FoldChange < -l,]
  return(list(up=rownames(up),down=rownames(down)))
}
## getAverage ##
# Get average expression (normalized by median of ratios) of each of the conditions in a deseq object.
# Taken from source code of the package deseqAbstraction which is no longer available on github.
# Credits to Per L. Brattås
# Parameters:
# dds = deseq object
getAverage <- function(dds) {
  baseMeanPerLvl <- sapply( levels(dds$condition), function(lvl) rowMeans( counts(dds,normalized=TRUE)[,dds$condition == lvl] ) )
  baseSDPerLvl <- sapply( levels(dds$condition), function(lvl) apply( counts(dds,normalized=TRUE)[,dds$condition == lvl],1,sd ) )
  colnames(baseSDPerLvl) <- paste("st.dev:",colnames(baseSDPerLvl),sep="")
  return(list(Mean=baseMeanPerLvl,SD=baseSDPerLvl))
}

meanPlot_cus <- function(exp,test,c1 = "condition 1",c2 = "condition 2",p=.05,l=0,id=F, ttl="", 
                         repel=TRUE, col1="tomato", col2="darkturquoise", col3="grey", highlights=NA){
  sign <- getSignName(x = test,p = p,l = l)
  u <- sign$up
  d <- sign$down
  
  #color up and down sign..
  colVec <- ifelse(test = (rownames(exp) %in% u),
                   yes = col1,
                   no = ifelse(test = (rownames(exp) %in% d),
                               yes = col2, no =col3))
  colVec[is.na(colVec)] <- "steelblue" ## if NA make sure it's not counted as <p
  #size of points
  cexVec <- ifelse(test = (rownames(exp) %in% u),
                   yes = 0.35,
                   no = ifelse(test = (rownames(exp) %in% d),
                               yes = 0.35, no = 0.3))
  
  exp_log <- as.data.frame(log2(exp[,c(c1, c2)] + 0.5))
  exp_log$Name <- rownames(exp_log)
  
  exp_log$reg <- factor(ifelse(exp_log$Name %in% u, paste('upregulated in ', c1, ' (', length(u), ')', sep =''),
                               ifelse(exp_log$Name %in% d, paste('downregulated in ', c1,' (', length(d), ')', sep =''), paste('not significant', ' (', (nrow(test) - length(u) - length(d)), ')', sep=''))))
  
  library(ggrepel)
  if(repel == TRUE){
    plt <- ggplot(exp_log, aes(x=get(c2), y=get(c1), label=Name, color=reg)) + geom_point(aes(size=cexVec), alpha=0.7)+ scale_color_manual(values=c(col2, col3, col1))+ scale_size_continuous(range=c(1,2), guide="none")+ geom_text_repel(data = subset(exp_log, Name %in% u | Name %in% d),direction    = "y", nudge_y = 0.4, nudge_x = -0.5)
  }
  else{
    plt <- ggplot(exp_log, aes(x=get(c2), y=get(c1), color=reg)) + geom_point(aes(size=cexVec), alpha=0.7)+ scale_color_manual(values=c(col2, col3, col1))+ scale_size_continuous(range=c(1,2), guide="none")
  }
  plt <- plt + labs(x=paste("log2(mean ",c2,")",sep=""), 
                    y=paste("log2(mean ",c1,")",sep=""),
                    title=paste(ttl, paste(c1," vs. ",c2,sep=""), sep = ': '),
                    subtitle=paste("p-adj < ",p,", log2(fc) > ",l,sep=""))+ theme_bw() +
    theme( plot.title = element_text( size=14, face="bold"), text = element_text(size=15), panel.grid.minor = element_blank(), panel.grid.major = element_blank(), legend.title=element_blank()) 
  
  
  if(id==T) {
    
    identify(log2(exp[,1]),log2(exp[,2]),labels = rownames(exp))
    
  }
  
  if(!is.na(highlights)){
    plt <- plt + geom_point(data=exp_log[highlights,], aes(x=get(c2), y=get(c1)), colour="springgreen3", size=5, shape=1, stroke=2)
  }
  return(plt)
  
}
```

## Loading gene count matrices

```{r}
library(data.table)
library(ggplot2)
library(ggpubr)
library(UpSetR)
library(openxlsx)
library(pheatmap)
library(stringr)
library(DESeq2)
library(bedr)

gene_annotation <- fread("/Volumes/MyPassport/annotations/human/gencode/v38/gencode.v38.annotation.gene_names.tab", data.table = F, skip=1, header = F)
colnames(gene_annotation) <- c("Geneid", "gene_name", "gene_type")

samplesheet <- fread("/Volumes/MyPassport/CRISPRi_L1s/bulk/samplesheet_L1crispr.tab", data.table = F, header = T)
samplesheet <- samplesheet[which(!samplesheet$samples %in% c("AA11_Sai2_lv3775_B_CRISPRa_S11", "AA_Sai2_L1_CRISPRa_LV3776_c_S29")),] # Outlier samples
samplesheet_list <- split(samplesheet, f = list(samplesheet$cell, samplesheet$experiment, samplesheet$guide, samplesheet$seqnum))
samplesheet_list <- samplesheet_list[sapply(samplesheet_list,nrow)>0] 

samplesheet_comparisons <- list(
  "hiPS6_crispri_g1" = rbind(samplesheet_list$hiPS6.crispri.g1.CTG_JGJSeq156_159_160truseq, 
                             samplesheet_list$hiPS6.crispri.LacZ.CTG_JGJSeq156_159_160truseq),
  "hiPS6_crispri_g2" = rbind(samplesheet_list$hiPS6.crispri.g2.CTG_JGJSeq205_207_208_214_truseq_2023_021,
                             samplesheet_list$hiPS6.crispri.LacZ.CTG_JGJSeq205_207_208_214_truseq_2023_021),
  "hiPS6_crispri_g3" = rbind(samplesheet_list$hiPS6.crispri.g3.CTG_JGJSeq205_207_208_214_truseq_2023_021,
                             samplesheet_list$hiPS6.crispri.LacZ.CTG_JGJSeq205_207_208_214_truseq_2023_021),
  "hiPS48_crispri_g3" = rbind(samplesheet_list$hiPS48.crispri.g3.testdata_ctg_2023_50,
                              samplesheet_list$hiPS48.crispri.LacZ.testdata_ctg_2023_50),
  "NES_crispra_g1" = rbind(samplesheet_list$NES.crispra.g1.CTG_JGJSeq156_159_160truseq,
                           samplesheet_list$NES.crispra.NTC.CTG_JGJSeq156_159_160truseq),
  "NES_crispra_g2" = rbind(samplesheet_list$NES.crispra.g2.CTG_JGJSeq233_237_250_257_2023_145,
                           samplesheet_list$NES.crispra.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "NES_crispra_g3" = rbind(samplesheet_list$NES.crispra.g3.CTG_JGJSeq233_237_250_257_2023_145,
                           samplesheet_list$NES.crispra.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "hiPS48_org_crispri_g3" = rbind(samplesheet_list$hiPS48_org.crispri.g3.CTGseq278_280_281_282_285_288_289_290_291_2023_282NovaseqX,
                                samplesheet_list$hiPS48_org.crispri.LacZ.CTGseq278_280_281_282_285_288_289_290_291_2023_282NovaseqX),
  "hiPS6_org_crispri_g3" = rbind(samplesheet_list$hiPS6_org.crispri.g3.CTG_JGJSeq233_237_250_257_2023_145,
                                  samplesheet_list$hiPS6_org.crispri.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "hiPS6_org_crispra_g3" = rbind(samplesheet_list$hiPS6_org.crispra.g3.CTG_JGJSeq233_237_250_257_2023_145,
                                  samplesheet_list$hiPS6_org.crispra.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "hiPS6_org_crispri_g1" = rbind(samplesheet_list$hiPS6_org.crispri.g1.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispri.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispri_g2" = rbind(samplesheet_list$hiPS6_org.crispri.g2.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispri.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispri_g3_batch2" = rbind(samplesheet_list$hiPS6_org.crispri.g3.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispri.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispra_g1" = rbind(samplesheet_list$hiPS6_org.crispra.g1.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispra.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispra_g2" = rbind(samplesheet_list$hiPS6_org.crispra.g2.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispra.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispra_g3_batch2" = rbind(samplesheet_list$hiPS6_org.crispra.g3.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispra.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq)
)


library(openxlsx)
write.xlsx(samplesheet_comparisons, "/Volumes/MyPassport/CRISPRi_L1s/results/tables/samplesheet_comparisons.xlsx")
path <- "/Volumes/MyPassport/CRISPRi_L1s/bulk/gene_counts/unique/"
# samples <- samplesheet$samples
samples <- unique(do.call(rbind, samplesheet_list))$samples
# Rerun featurecounts to the organoids 
for(i in 1:length(samples)){
  sample <- samples[i]
  if(i == 1){
    gene_counts <- fread(paste(path, sample, "_gene_count_matrix_2.csv", sep=""), data.table = F)    
    colnames(gene_counts)[ncol(gene_counts)] <- sample
    rownames(gene_counts) <- gene_counts$Geneid
    row_order <- rownames(gene_counts)
  }else{
    tmp <- fread(paste(path, sample, "_gene_count_matrix_2.csv", sep=""), data.table = F)
    colnames(tmp)[ncol(tmp)] <- sample
    rownames(tmp) <- tmp$Geneid
    gene_counts <- cbind(gene_counts[row_order,], tmp[row_order,sample,drop=F])
  }
}
```

## Cell type markers

Normalization using sizeFactors calculated with all samples. 
```{r}
rownames(samplesheet) <- samplesheet$samples
rownames(gene_counts) <- (gene_counts$Geneid)
gene_dds <- DESeqDataSetFromMatrix(gene_counts[,rownames(samplesheet)], samplesheet, design =  ~ condition+seqnum)
gene_dds$condition <- relevel(gene_dds$condition, "Control")
gene_dds <- DESeq(gene_dds)

# write.table(as.data.frame(gene_dds$sizeFactor), file = "/Volumes/MyPassport/CRISPRi_L1s/results/tables/gene_sizeFactors.tab", quote = F, sep="\t", col.names = F)

gene_count_norm <- counts(gene_dds, normalized=T)
neuronal_genes <- c("WNT3", "POU3F2", "NEUROD4", "NEUROG2", "FOXG1", "NEUROD1", "SOX11", "NES", "SOX1", "SOX2", "ASCL1", "DCX", "NCAM1", "MAP2")
ectoderm <- c("HES5", "PAMR1", "PAX6")
endoderm <- c("CER1", "EOMES", "GATA6")
mesoderm <- c("APLNR", "HAND1", "HOXB7")
trophectoderm <- c("GATA3", "PTGES", "PDGFA")
pluripotency_genes <- c("POU5F1", "NANOG")

samplesheet <- samplesheet[order(samplesheet$cell, samplesheet$LV),]
samplesheet_hips48 <- samplesheet[which(samplesheet$cell == "hiPS48"),]
samplesheet_hips6 <- samplesheet[which(samplesheet$cell == "hiPS6"),]
samplesheet_hips6 <- samplesheet_hips6[order(samplesheet_hips6$condition),]
samplesheet_hips6_batch1 <- samplesheet_hips6[which(samplesheet_hips6$seqnum == "CTG_JGJSeq156_159_160truseq"),]
samplesheet_hips6_batch2 <- samplesheet_hips6[which(samplesheet_hips6$seqnum == "CTG_JGJSeq205_207_208_214_truseq_2023_021"),]

markers <- data.frame(gene = c("neuronal" = neuronal_genes, 
                        "pluripotency" = pluripotency_genes, 
                        "ectoderm" = ectoderm, 
                        "endoderm" = endoderm, 
                        "mesoderm" = mesoderm, 
                        "trophectoderm" = trophectoderm))
markers$type <- gsub("^\\d+|\\d+$", "", rownames(markers))    
rownames(markers) <- markers$gene


# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/markers_heatmaps.pdf", height = 10, width = 15)
pheatmap(log2(gene_count_norm[c(neuronal_genes, pluripotency_genes),rownames(samplesheet)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(16), main = "All samples", annotation_col = samplesheet[,c("LV", "cell")])
pheatmap(log2(gene_count_norm[markers$gene,rownames(samplesheet)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(16), main = "All samples", annotation_row = markers[,"type", drop=F], annotation_col = samplesheet[,c("LV", "cell")])
pheatmap(log2(gene_count_norm[c(neuronal_genes, pluripotency_genes),rownames(samplesheet_hips6_batch1)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(3), main = "hiPS6 Batch 1", annotation_col = samplesheet[,c("LV", "cell")])
pheatmap(log2(gene_count_norm[markers$gene,rownames(samplesheet_hips6_batch1)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(3), main = "hiPS6 Batch 1", annotation_row = markers[,"type", drop=F], annotation_col = samplesheet[,c("LV", "cell")])

pheatmap(log2(gene_count_norm[c(neuronal_genes, pluripotency_genes),rownames(samplesheet_hips6_batch2)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(3,6), main = "hiPS6 Batch 2", annotation_col = samplesheet[,c("LV", "cell")])
pheatmap(log2(gene_count_norm[markers$gene,rownames(samplesheet_hips6_batch2)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(3,6), main = "hiPS6 Batch 2", annotation_row = markers[,"type", drop=F], annotation_col = samplesheet[,c("LV", "cell")])

pheatmap(log2(gene_count_norm[c(neuronal_genes, pluripotency_genes),rownames(samplesheet_hips48)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(3), main = "hiPS48", annotation_col = samplesheet[,c("LV", "cell")])
pheatmap(log2(gene_count_norm[markers$gene,rownames(samplesheet_hips48)]+0.5), cluster_cols = F, cluster_rows = F, scale = "none", gaps_col = c(3), main = "hiPS48", annotation_row = markers[,"type", drop=F], annotation_col = samplesheet[,c("LV", "cell")])

# dev.off()

```

## Gene Differential Expression Analysis (DEA)

Statistical analysis per experiment using DESeq2.
```{r}
gene_dds_list <- list()
gene_res_list <- list()
gene_res_list_df <- list()
gene_exp_list <- list()
rownames(gene_counts) <- gene_counts$Geneid

for(i in names(samplesheet_comparisons)){
  rownames(samplesheet_comparisons[[i]]) <- samplesheet_comparisons[[i]]$samples
  print(i)
  print(c("Batch: ", unique(samplesheet_comparisons[[i]]$seqnum)))
  conditions <- unique(samplesheet_comparisons[[i]]$condition)
  print(c("Conditions: ", conditions))

  effect <- conditions[which(conditions != "Control")]
  test <- paste("condition_", effect ,"_vs_Control", sep = "")
  print(c("Test: ", test))
  counts <- gene_counts[,rownames(samplesheet_comparisons[[i]])]
  gene_dds_list[[i]] <- DESeqDataSetFromMatrix(counts, samplesheet_comparisons[[i]], design =  ~ condition)
  gene_dds_list[[i]]$condition <- relevel(gene_dds_list[[i]]$condition, "Control")
  gene_dds_list[[i]] <- DESeq(gene_dds_list[[i]])
  gene_res_list[[i]] <- lfcShrink(gene_dds_list[[i]], test)
  gene_exp_list[[i]] <- getAverage(gene_dds_list[[i]])
  gene_res_list_df[[i]] <- as.data.frame(gene_res_list[[i]])
  gene_res_list_df[[i]]$gene_name <- rownames(gene_res_list_df[[i]])
}

```

## PCA plots
```{r}
gene_vst_list <- list()
for (i in names(gene_dds_list)) gene_vst_list[[i]] <- varianceStabilizingTransformation(gene_dds_list[[i]])

gene_vst <- varianceStabilizingTransformation(gene_dds)
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/gene_PCAs.pdf")
plotPCA(gene_vst, intgroup="seqnum") + ggtitle("All samples: Batch") + theme_bw()
plotPCA(gene_vst, intgroup="cell") + ggtitle("All samples: Cell type/line") + theme_bw()
plotPCA(gene_vst, intgroup="experiment") + ggtitle("All samples: Experiment") + theme_bw()
plotPCA(gene_vst, intgroup="condition") + ggtitle("All samples: Condition") + theme_bw()
plotPCA(gene_vst_list$hiPS6_crispri_g1) + ggtitle("hiPS6_crispri_g1") + theme_bw()
plotPCA(gene_vst_list$hiPS6_crispri_g3) + ggtitle("hiPS6_crispri_g3") + theme_bw()
plotPCA(gene_vst_list$hiPS48_crispri_g3) + ggtitle("hiPS48_crispri_g3") + theme_bw()
plotPCA(gene_vst_list$hiPS48_org_crispri_g3) + ggtitle("hiPS48_org_crispri_g3") + theme_bw()
plotPCA(gene_vst_list$hiPS6_org_crispri_g3) + ggtitle("hiPS6_org_crispri_g3") + theme_bw()
plotPCA(gene_vst_list$hiPS6_org_crispri_g3_batch2) + ggtitle("hiPS6_org_crispri_g3_batch2") + theme_bw()
plotPCA(gene_vst_list$hiPS6_org_crispri_g1) + ggtitle("hiPS6_org_crispri_g1") + theme_bw()
# dev.off()
```

## Visualization

Mean plots, volcano plots and excel tables...
```{r}
library(EnhancedVolcano)
gene_annotation <- fread("/Volumes/MyPassport/annotations/human/repeatmasker/hg38_rmsk_TEtranscripts_classification.tab", data.table = F, header = F)
colnames(gene_annotation) <- c("gene_id", "gene_subfamily", "gene_family", "gene_class")

volcano_plots_list <- list()
mean_plots_list <- list()
for(i in names(gene_res_list)){
  tmp <- EnhancedVolcano(gene_res_list[[i]], 
                                             drawConnectors = T, 
                                             lab = NA, 
                                             x = 'log2FoldChange',
                                             y = 'padj',
                                             pCutoff = 0.05,
                                             FCcutoff = 1)
  tmp$data$Sig <- ifelse(tmp$data$log2FoldChange > 1 & tmp$data$Sig == "FC_P", "Upregulated", 
                                             ifelse(tmp$data$log2FoldChange < -1 & tmp$data$Sig == "FC_P", "Downregulated", "Not significant"))
  tmp$data$Sig <- paste(tmp$data$Sig, table(tmp$data$Sig)[tmp$data$Sig], sep = ": ")
  
  not_sig_lab <- unique(tmp$data$Sig)[startsWith(unique(tmp$data$Sig), "Not significant")]
  up_sig_lab <- unique(tmp$data$Sig)[startsWith(unique(tmp$data$Sig), "Upregulated")]
  down_sig_lab <- unique(tmp$data$Sig)[startsWith(unique(tmp$data$Sig), "Downregulated")]
  
  point_colours <- c("grey","darkturquoise", "tomato")
  names(point_colours) <- c(not_sig_lab, down_sig_lab, up_sig_lab)
  point_sizes <- c(1,2,2)
  names(point_sizes) <- c(not_sig_lab, down_sig_lab, up_sig_lab)  
  
  title <- str_replace_all(i, pattern = "_", replacement = " ")
  
  tmp <- tmp$data
  volcano_plots_list[[i]] <- ggplot(tmp, aes(x=log2FoldChange, y=-log10(padj), colour=Sig, size=Sig)) + geom_point(alpha=0.7) + 
    geom_vline(xintercept = 1, linetype="dotted") + geom_vline(xintercept = -1, linetype="dotted") + 
    geom_hline(yintercept = -log10(0.05), linetype="dotted") + labs(colour="") + 
    scale_color_manual(values = point_colours)  + 
    scale_size_manual(values = point_sizes, guide="none") +
    theme_bw() +
    theme(text = element_text(size=15), panel.grid.minor = element_blank(), panel.grid.major = element_blank()) +
    ggtitle(title)
  
  effect <- unique(samplesheet_comparisons[[i]][which(samplesheet_comparisons[[i]]$condition != "Control"),"condition"])
  mean_plots_list[[i]] <- meanPlot_cus(exp = gene_exp_list[[i]]$Mean, 
                                       test = gene_res_list[[i]], 
                                       c1 = effect, c2="Control", p = 0.05, l=1, ttl = title, repel = F)
    
}

order <- c("hiPS6_crispri_g1",
"hiPS6_crispri_g3",
"hiPS48_crispri_g3",
"hiPS6_org_crispri_g1",
"hiPS48_org_crispri_g3",
"hiPS6_org_crispri_g3_batch2",
"hiPS6_org_crispri_g3")

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/genes_volcano_plots.pdf", height = 4, width = 7)
volcano_plots_list
# dev.off()

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/genes_mean_plots.pdf", height = 4, width = 7)
# mean_plots_list
# dev.off()
```

## Gene Set Enrichment Analysis

Dot plots from top terms (and neuronal related) + excel tables 
```{r}
library(clusterProfiler)
library(org.Hs.eg.db)
set.seed(10)
gse_list <- list()
for(i in names(gene_res_list_df)){
  genelist <- gene_res_list_df[[i]][,c("log2FoldChange"), drop=F]
  genelist$gene_name <- rownames(genelist)
  genelist <- genelist[order(genelist$log2FoldChange, decreasing = T),]
  genelist_FC <- genelist$log2FoldChange
  names(genelist_FC) <- genelist$gene_name
  
  gse_list[[i]] <- gseGO(geneList=sort(genelist_FC, decreasing = T), 
               ont ="BP", 
               keyType = "SYMBOL", 
               minGSSize = 3, 
               maxGSSize = 800, 
               pvalueCutoff = 0.05,
               verbose = TRUE, 
               seed = TRUE,
               OrgDb = org.Hs.eg.db, 
               pAdjustMethod = "BH")
}

library(dplyr)
top20_gsea_dotplot <- function(gse, name){
  gse@result$sign <- ifelse(gse@result$enrichmentScore > 0, "Activated", "Supressed")
  gse@result$num_genes <- sapply(str_split(gse@result$core_enrichment, "/"), length)
  gse@result$gene_ratio <- gse@result$num_genes / gse@result$setSize
  
  gse@result <- top_n(gse@result, -20, gse@result$p.adjust) %>% arrange(gene_ratio)
  if(nrow(gse@result) > 20){
    gse@result <- top_n(gse@result, 20, abs(gse@result$enrichmentScore)) %>% arrange(gene_ratio)
  }
  gse@result$Description <- factor(gse@result$Description, levels = unique(gse@result$Description))
  
  ggplot(gse@result, aes(x=Description, y=gene_ratio, size = num_genes, fill = p.adjust)) + geom_point(shape=21, colour="lightgrey") + facet_wrap(.~sign) + coord_flip() + theme_pubr(legend = "right", border = T) + labs(x="", fill="Padj", y = "Gene ratio", size = "Num genes") + scale_fill_gradientn(colors = c("firebrick1", "gray94")) + scale_size_continuous(range = c(2,8)) + 
    theme(axis.text.y = element_text(size=10), strip.text.x = element_text(size = 10),axis.text.x = element_text(size=10),legend.text = element_text(size=10),legend.title = element_text(size=10)) + 
    ggtitle(paste(name, ":\nTop 20 most significant"))
}

top20_gsea <- list()
for( g in names(gse_list)){
  print(g)
  print(nrow(gse_list[[g]]@result))
  if(nrow(gse_list[[g]]@result) > 0){
    top20_gsea[[g]] <- top20_gsea_dotplot(gse = gse_list[[g]], name=g)  
  }
}
# dir.create("/Volumes/MyPassport/CRISPRi_L1s/results/plots/GSEA_top20_padj")
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/GSEA_top20_padj/hiPS6_org_crispri_g1_gsea.pdf", width = 7)
top20_gsea$hiPS6_org_crispri_g1
# dev.off()
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/GSEA_top20_padj/hiPS6_org_crispri_g3_gsea.pdf", width = 8)
top20_gsea$hiPS6_org_crispri_g3
# dev.off()
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/GSEA_top20_padj/hiPS6_org_crispri_g3_batch2_gsea.pdf", width = 7)
top20_gsea$hiPS6_org_crispri_g3_batch2
# dev.off()
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/GSEA_top20_padj/hiPS48_org_crispri_g3_gsea.pdf", width = 8)
top20_gsea$hiPS48_org_crispri_g3
# dev.off()

gse_list_df <- list("hiPS6_crispri_g1" = gse_list$hiPS6_crispri_g1@result,
                    "hiPS6_crispri_g3" = gse_list$hiPS6_crispri_g3@result,
                    "hiPS6_org_crispri_g1" = gse_list$hiPS6_org_crispri_g1@result,
                    "hiPS6_org_crispri_g3" = gse_list$hiPS6_org_crispri_g3@result,
                    "hiPS6_org_crispri_g3_batch2" = gse_list$hiPS6_org_crispri_g3_batch2@result)

for(i in names(gse_list_df)){
  gse_list_df[[i]]$sign <- ifelse(gse_list_df[[i]]$enrichmentScore > 0, "Activated", "Supressed")
  gse_list_df[[i]]$num_genes <- sapply(str_split(gse_list_df[[i]]$core_enrichment, "/"), length)
  gse_list_df[[i]]$gene_ratio <- gse_list_df[[i]]$num_genes / gse_list_df[[i]]$setSize
  
  rownames(gse_list_df[[i]]) <- gse_list_df[[i]]$Description
  gse_list_df[[i]]$Description <- factor(gse_list_df[[i]]$Description, levels= gse_list_df[[i]][order(gse_list_df[[i]]$gene_ratio, decreasing = F),"Description"])
}

openxlsx::write.xlsx(gse_list_df, "/Volumes/MyPassport/CRISPRi_L1s/bulk/results/tables/genes_GSEA_crispri.xlsx")

```

## Overlap FL L1HS-PA4 genes' TSS

Gene TSS in windows of 1K, 2K, 5K, 10K, 15K, 20K, 50K bp of FL-L1PAs. Focus only on protein coding genes.
```{r}
FL_L1PAs <- fread("/Volumes/MyPassport/CRISPRi_L1s/results/tables/hg38_rmsk_6kbp_l1hs_l1pa2_l1pa3_l1pa4.bed", data.table=F, skip = 1)
colnames(FL_L1PAs) <- c("chr", "start", "end", "TE_id", "dot", "strand", "TE_info")

add_windows_coords <- function(df, window = 0){
  if(window > 0){
    for(i in 1:nrow(df)){
      df[i,"start"] <- df[i,"start"] - window
      df[i,"start"] <- ifelse(df[i,"start"] <= 0, 1, df[i,"start"])
      df[i,"end"] <- df[i,"end"] + window
    }
  }
  tmp <- data.frame(id = rownames(df), coords = str_replace_all(paste(df$chr, paste(format(df$start, scientific = F), format(df$end, scientific = F), sep = "-"), sep = ":"), pattern = " ", replacement = ""))
  tmp <- tmp[is.valid.region(tmp$coords),]
  return(tmp)
}

FL_L1PAs <- FL_L1PAs[!grepl("_", FL_L1PAs$chr),]
FL_L1PAs$coords <- add_windows_coords(FL_L1PAs)$coords
FL_L1PAs$coords_1kb <- add_windows_coords(FL_L1PAs, window = 1000)$coords
FL_L1PAs$coords_2kb <- add_windows_coords(FL_L1PAs, window = 2000)$coords
FL_L1PAs$coords_5kb <- add_windows_coords(FL_L1PAs, window = 5000)$coords
FL_L1PAs$coords_10kb <- add_windows_coords(FL_L1PAs, window = 10000)$coords
FL_L1PAs$coords_15kb <- add_windows_coords(FL_L1PAs, window = 15000)$coords
FL_L1PAs$coords_20kb <- add_windows_coords(FL_L1PAs, window = 20000)$coords
FL_L1PAs$coords_50kb <- add_windows_coords(FL_L1PAs, window = 50000)$coords

# BED file with gene transcription start sites
gene_bed <- fread("/Volumes/MyPassport/annotations/human/gencode/v38/gencode.v38.annotation.transcripts.tss_plus5.bed", data.table=F, skip=1, header=F)
gene_type <- fread("/Volumes/MyPassport/annotations/human/gencode/v38/gencode.v38.annotation.gene_names.tab", data.table=F, skip=1, header=F)
colnames(gene_bed) <- c("chr", "start", "end", "dot", "strand", "dot1", "gene_id", "transcript_id", "gene_name")
colnames(gene_type) <- c("gene_id", "gene_name", "gene_type")
gene_bed <- merge(gene_bed, gene_type, by=c("gene_id", "gene_name"))
gene_bed <- gene_bed[!grepl("_", gene_bed$chr),]

gene_bed$coords <- add_windows_coords(gene_bed)$coords
gene_bed <- gene_bed[match(bedr.sort.region(gene_bed$coords), gene_bed$coords),]
gene_bed <- gene_bed[which(gene_bed$gene_type == "protein_coding"),] # Focus only on protein coding genes
rownames(gene_bed) <- make.unique(gene_bed$gene_name)
gene_res_list_df_protein <- gene_res_list_df
for(i in names(gene_res_list_df_protein)) gene_res_list_df_protein[[i]] <- gene_res_list_df[[i]][which(rownames(gene_res_list_df[[i]]) %in% rownames(gene_bed)),]

FL_L1PAs <- FL_L1PAs[match(bedr.sort.region(FL_L1PAs$coords), FL_L1PAs$coords),]

# intersect
genes_TSS_within_1kb_FL_L1PAs <- gene_bed[which(in.region(x = gene_bed$coords, y = FL_L1PAs$coords_1kb)), c("gene_name", "transcript_id", "coords")]
genes_TSS_within_2kb_FL_L1PAs <- gene_bed[which(in.region(x = gene_bed$coords, y = FL_L1PAs$coords_2kb)), c("gene_name", "transcript_id", "coords")]
genes_TSS_within_5kb_FL_L1PAs <- gene_bed[which(in.region(x = gene_bed$coords, y = FL_L1PAs$coords_5kb)), c("gene_name", "transcript_id", "coords")]
genes_TSS_within_10kb_FL_L1PAs <- gene_bed[which(in.region(x = gene_bed$coords, y = FL_L1PAs$coords_10kb)), c("gene_name", "transcript_id", "coords")]
genes_TSS_within_15kb_FL_L1PAs <- gene_bed[which(in.region(x = gene_bed$coords, y = FL_L1PAs$coords_15kb)), c("gene_name", "transcript_id", "coords")]
genes_TSS_within_20kb_FL_L1PAs <- gene_bed[which(in.region(x = gene_bed$coords, y = FL_L1PAs$coords_20kb)), c("gene_name", "transcript_id", "coords")]
genes_TSS_within_50kb_FL_L1PAs <- gene_bed[which(in.region(x = gene_bed$coords, y = FL_L1PAs$coords_50kb)), c("gene_name", "transcript_id", "coords")]

genes_TSS_nearby_FL_L1PAs <- list()

# Removing duplicate genes within groups
genes_TSS_nearby_FL_L1PAs[["1kb"]] <- genes_TSS_within_1kb_FL_L1PAs
genes_TSS_nearby_FL_L1PAs[["2kb"]] <- genes_TSS_within_2kb_FL_L1PAs[which(!genes_TSS_within_2kb_FL_L1PAs$gene_name %in% genes_TSS_within_1kb_FL_L1PAs$gene_name),]
genes_TSS_nearby_FL_L1PAs[["5kb"]] <- genes_TSS_within_5kb_FL_L1PAs[which(!genes_TSS_within_5kb_FL_L1PAs$gene_name %in% c(genes_TSS_within_2kb_FL_L1PAs$gene_name, genes_TSS_within_1kb_FL_L1PAs$gene_name)),]
genes_TSS_nearby_FL_L1PAs[["10kb"]] <- genes_TSS_within_10kb_FL_L1PAs[which(!genes_TSS_within_10kb_FL_L1PAs$gene_name %in% c(genes_TSS_within_5kb_FL_L1PAs$gene_name, genes_TSS_within_2kb_FL_L1PAs$gene_name, genes_TSS_within_1kb_FL_L1PAs$gene_name)),]
genes_TSS_nearby_FL_L1PAs[["15kb"]] <- genes_TSS_within_15kb_FL_L1PAs[which(!genes_TSS_within_15kb_FL_L1PAs$gene_name %in% c(genes_TSS_within_10kb_FL_L1PAs$gene_name, genes_TSS_within_5kb_FL_L1PAs$gene_name, genes_TSS_within_2kb_FL_L1PAs$gene_name, genes_TSS_within_1kb_FL_L1PAs$gene_name)),]
genes_TSS_nearby_FL_L1PAs[["20kb"]] <- genes_TSS_within_20kb_FL_L1PAs[which(!genes_TSS_within_20kb_FL_L1PAs$gene_name %in% c(genes_TSS_within_15kb_FL_L1PAs$gene_name, genes_TSS_within_10kb_FL_L1PAs$gene_name, genes_TSS_within_5kb_FL_L1PAs$gene_name, genes_TSS_within_2kb_FL_L1PAs$gene_name, genes_TSS_within_1kb_FL_L1PAs$gene_name)),]
genes_TSS_nearby_FL_L1PAs[["50kb"]] <- genes_TSS_within_50kb_FL_L1PAs[which(!genes_TSS_within_50kb_FL_L1PAs$gene_name %in% c(genes_TSS_within_20kb_FL_L1PAs$gene_name, genes_TSS_within_15kb_FL_L1PAs$gene_name, genes_TSS_within_10kb_FL_L1PAs$gene_name, genes_TSS_within_5kb_FL_L1PAs$gene_name, genes_TSS_within_2kb_FL_L1PAs$gene_name, genes_TSS_within_1kb_FL_L1PAs$gene_name)),]

for(j in names(gene_res_list_df_protein)){
  for(i in names(genes_TSS_nearby_FL_L1PAs)) gene_res_list_df_protein[[j]][genes_TSS_nearby_FL_L1PAs[[i]]$gene_name,"FL_L1PAs"] <- i
  gene_res_list_df_protein[[j]]$FL_L1PAs <- factor(gene_res_list_df_protein[[j]]$FL_L1PAs, levels = names(genes_TSS_nearby_FL_L1PAs))
}

# Identify protein coding genes affected by L1-CRISPRi (log2FC > 1) nearby FL-L1s
downreg_hiPS48_g3_FL_L1PAs_TSS <- gene_res_list_df_protein$hiPS48_crispri_g3[which(!is.na(gene_res_list_df_protein$hiPS48_crispri_g3$FL_L1PAs) & gene_res_list_df_protein$hiPS48_crispri_g3$log2FoldChange < -1),]
downreg_hiPS6_g3_FL_L1PAs_TSS <- gene_res_list_df_protein$hiPS6_crispri_g3[which(!is.na(gene_res_list_df_protein$hiPS6_crispri_g3$FL_L1PAs) & gene_res_list_df_protein$hiPS6_crispri_g3$log2FoldChange < -1),]
downreg_hiPS6_g1_FL_L1PAs_TSS <- gene_res_list_df_protein$hiPS6_crispri_g1[which(!is.na(gene_res_list_df_protein$hiPS6_crispri_g1$FL_L1PAs) & gene_res_list_df_protein$hiPS6_crispri_g1$log2FoldChange < -1),]

# save.image("/Volumes/MyPassport/CRISPRi_L1s/src/r_scripts/gene_uniq_DEA.Rdata")
```

Coordinates to the downregulated genes and how close they are to a FL-L1 (Fig 4b).
```{r}
downreg_gene_hiPS_crispri <- unique(gene_res_list_df$hiPS48_crispri_g3[which(gene_res_list_df$hiPS48_crispri_g3$padj < 0.05 & gene_res_list_df$hiPS48_crispri_g3$log2FoldChange < -1), "gene_name"])

# Reload annotation files, I need the coordinates to the downregulated genes and how close they are to a FL-L1 
gene_type <- fread("/Volumes/MyPassport/annotations/human/gencode/v38/gencode.v38.annotation.gene_names.tab", skip=1, header = F, data.table = F)
colnames(gene_type) <- c("gene_id", "gene_name", "gene_type")
gene_type <- unique(gene_type[,2:3])
rownames(gene_type) <- make.unique(gene_type$gene_name)

downreg_gene_hiPS_crispri_type <- gene_type[downreg_gene_hiPS_crispri,]
downreg_gene_hiPS_crispri_type$gene_type <- ifelse(downreg_gene_hiPS_crispri_type$gene_type != "protein_coding", "ncRNA", downreg_gene_hiPS_crispri_type$gene_type)
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/pie_downreg_crispri_hiPS48_g3.pdf", height = 4, width = 4)
pie(table(downreg_gene_hiPS_crispri_type$gene_type))
# dev.off()

gene_bed <- fread("/Volumes/MyPassport/annotations/human/gencode/v38/gencode.v38.annotation.bed", data.table=F, skip=1, header=F)
colnames(gene_bed) <- c("chr", "start", "end", "gene_id", "dot", "strand", "gene_name")
rownames(gene_bed) <- make.unique(gene_bed$gene_name)
downreg_gene_hiPS_crispri_bed <- gene_bed[downreg_gene_hiPS_crispri,]
downreg_gene_hiPS_crispri_bed$coords <- add_windows_coords(downreg_gene_hiPS_crispri_bed)$coords
downreg_gene_hiPS_crispri_bed <- downreg_gene_hiPS_crispri_bed[match(bedr.sort.region(downreg_gene_hiPS_crispri_bed$coords), downreg_gene_hiPS_crispri_bed$coords),]

# write.table(downreg_gene_hiPS_crispri_bed[,c(1:3, 7, 5,6)], "/Volumes/MyPassport/CRISPRi_L1s/results/tables/downreg_gene_hiPS48_g3_crispri.bed", sep="\t", row.names = F, col.names = F, quote = F)

# downreg_gene_hiPS_crispri_intersect <- fread("/Volumes/MyPassport/CRISPRi_L1s/results/tables/downreg_gene_hiPS_crispri_intersect_6kbp_l1hs_pa4.bed", skip=1, header = F, data.table = F)

downreg_gene_hiPS_crispri_intersect_table <- data.frame(gene_name = downreg_gene_hiPS_crispri_type$gene_name,
           gene_type = downreg_gene_hiPS_crispri_type$gene_type,
           intersect_L1HS_PA4 = ifelse(downreg_gene_hiPS_crispri_type$gene_name %in% downreg_gene_hiPS_crispri_intersect$V4, T, F))

downreg_gene_hiPS_crispri_intersect_table <- as.data.frame(table(downreg_gene_hiPS_crispri_intersect_table$intersect_L1HS_PA4, downreg_gene_hiPS_crispri_intersect_table$gene_type))
colnames(downreg_gene_hiPS_crispri_intersect_table) <- c("intersect_L1HS_PA4", "type", "Count")
downreg_gene_hiPS_crispri_intersect_table$type <- factor(downreg_gene_hiPS_crispri_intersect_table$type, levels = c("lncRNA", "protein_coding", "ncRNA"))
downreg_gene_hiPS_crispri_intersect_table$intersect_L1HS_PA4 <- factor(downreg_gene_hiPS_crispri_intersect_table$intersect_L1HS_PA4, levels = c("TRUE","FALSE"))
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/bar_downreg_crispri_hiPS48_g3_intersect_L1.pdf", height = 3, width = 5)
ggplot(downreg_gene_hiPS_crispri_intersect_table[which(downreg_gene_hiPS_crispri_intersect_table$intersect_L1HS_PA4 == "TRUE"), ], aes(x=type, y=Count, fill=intersect_L1HS_PA4)) + geom_bar(stat = "identity") + theme_bw()
ggplot(downreg_gene_hiPS_crispri_intersect_table, aes(x=type, y=Count, fill=intersect_L1HS_PA4)) + geom_bar(stat = "identity") + theme_bw()
ggplot(downreg_gene_hiPS_crispri_intersect_table, aes(x=type, y=Count, fill=intersect_L1HS_PA4)) + geom_bar(stat = "identity", position = "fill") + theme_bw()
# dev.off()
```

## Proteomics

Check if we can see this effect on a protein level
```{r}
proteomics <- read.xlsx("/Volumes/MyPassport/CRISPRi_L1s/results/tables/proteomics_L1_CRISPRi_240417.xlsx")
colnames(proteomics)
rownames(proteomics) <- proteomics$Gene.symbol
samples_proteomics <- c("Log2.gRNA1-1","Log2.gRNA1-2", "Log2.gRNA1-3", "Log2.gRNA2-1", "Log2.gRNA2-2", "Log2.gRNA2-3", "Log2.gRNA3-1", "Log2.gRNA3-2", "Log2.gRNA3-3", "Log2.LacZ-1", "Log2.LacZ-2", "Log2.LacZ-3", "log2FC.gRNA1-LacZ", "log2FC.gRNA2-LacZ", "log2FC.gRNA3-LacZ")
for(i in samples_proteomics) proteomics[,i] <- as.numeric(as.character(proteomics[,i]))
samples_proteomics <- c("Log2.gRNA1-1","Log2.gRNA1-2", "Log2.gRNA1-3", "Log2.gRNA2-1", "Log2.gRNA2-2", "Log2.gRNA2-3", "Log2.gRNA3-1", "Log2.gRNA3-2", "Log2.gRNA3-3", "Log2.LacZ-1", "Log2.LacZ-2", "Log2.LacZ-3")

coldata <- data.frame(samples = samples_proteomics)
coldata$condition <- ifelse(grepl("gRNA1", coldata$samples), "g1",
                            ifelse(grepl("gRNA2", coldata$samples), "g2",
                                   ifelse(grepl("gRNA3", coldata$samples), "g3", 
                                          ifelse(grepl("LacZ", coldata$samples), "LacZ", "Other?"))))
coldata$condition <- factor(coldata$condition, levels = c("LacZ", "g1", "g2", "g3"))
coldata <- coldata[!grepl("gRNA2", coldata$samples),]
coldata$samples <- factor(coldata$samples, levels = coldata[order(coldata$condition),"samples"])
rownames(coldata) <- coldata$samples

downreg_hiPS48_org_crispri_g3_FL_L1PAs_TSS$type <- "hiPS48_org_crispri_g3"
downreg_hiPS6_org_crispri_g3_batch2_FL_L1PAs_TSS$type <- "hiPS6_org_crispri_g3_batch2"
downreg_hiPS6_org_crispri_g3_FL_L1PAs_TSS$type <- "hiPS6_org_crispri_g3"
downreg_hiPS6_org_crispri_g1_FL_L1PAs_TSS$type <- "hiPS6_org_crispri_g1"
downreg_hiPS48_g3_FL_L1PAs_TSS$type <- "hiPS48_g3"
downreg_hiPS6_g3_FL_L1PAs_TSS$type <- "hiPS6_g3"
downreg_hiPS6_g1_FL_L1PAs_TSS$type <- "hiPS6_g1"
all_downreg <- rbind(downreg_hiPS48_org_crispri_g3_FL_L1PAs_TSS, # All interesting genes
                     downreg_hiPS6_org_crispri_g3_batch2_FL_L1PAs_TSS,
                     downreg_hiPS6_org_crispri_g3_FL_L1PAs_TSS,
                     downreg_hiPS6_org_crispri_g1_FL_L1PAs_TSS,
                     downreg_hiPS48_g3_FL_L1PAs_TSS,
                     downreg_hiPS6_g3_FL_L1PAs_TSS,
                     downreg_hiPS6_g1_FL_L1PAs_TSS)
library(dplyr)
all_downreg <- all_downreg[,c("gene_name", "type")] %>% group_by_at(vars(gene_name)) %>%
  summarize_all(paste, collapse=",") %>% as.data.frame()
rownames(all_downreg) <- all_downreg$gene_name

tmp <- proteomics[unique(c("L1RE1", all_downreg[which(all_downreg$gene_name %in% proteomics$Gene.symbol),"gene_name"])),c("log2FC.gRNA1-LacZ", "log2FC.gRNA2-LacZ", "log2FC.gRNA3-LacZ")]
rg <- max(abs(tmp[!is.na(tmp)]));
proteomics[is.na(proteomics$`log2FC.gRNA1-LacZ`), "log2FC.gRNA1-LacZ"] <- 0
proteomics[is.na(proteomics$`log2FC.gRNA2-LacZ`), "log2FC.gRNA2-LacZ"] <- 0
proteomics[is.na(proteomics$`log2FC.gRNA3-LacZ`), "log2FC.gRNA3-LacZ"] <- 0
proteomics[is.na(proteomics$`Log2.gRNA1-1`), "Log2.gRNA1-1"] <- 0
proteomics[is.na(proteomics$`Log2.gRNA1-2`), "Log2.gRNA1-2"] <- 0
proteomics[is.na(proteomics$`Log2.gRNA1-3`), "Log2.gRNA1-3"] <- 0
proteomics[is.na(proteomics$`Log2.gRNA3-1`), "Log2.gRNA3-1"] <- 0
proteomics[is.na(proteomics$`Log2.gRNA3-2`), "Log2.gRNA3-2"] <- 0
proteomics[is.na(proteomics$`Log2.gRNA3-3`), "Log2.gRNA3-3"] <- 0
proteomics[is.na(proteomics$`Log2.LacZ-1`), "Log2.LacZ-1"] <- 0
proteomics[is.na(proteomics$`Log2.LacZ-2`), "Log2.LacZ-2"] <- 0
proteomics[is.na(proteomics$`Log2.LacZ-3`), "Log2.LacZ-3"] <- 0

downreg_genes_proteins_nearby_L1 <- unique(all_downreg[which(all_downreg$gene_name %in% proteomics$Gene.symbol),"gene_name"])
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/heatmap_downreg_nearby_FL_L1PAs_log2FC_proteomics.pdf", height = 20, width = 12)
pheatmap(proteomics[c(downreg_genes_proteins_nearby_L1, "L1RE1"),c("log2FC.gRNA1-LacZ", "log2FC.gRNA3-LacZ")], color=c(colorRampPalette(c( "navy", "lightskyblue", "mintcream"))(50), colorRampPalette(c("mistyrose", "tomato", "darkred"))(50)), cluster_rows = T, cluster_cols = F, breaks = seq(-rg, rg, length.out = 100), border_color = NA, annotation_row = all_downreg[downreg_genes_proteins_nearby_L1,"type", drop=F])
# dev.off()

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/heatmap_downreg_nearby_FL_L1PAs_proteomics.pdf", height = 20, width = 12)
tmp <- pheatmap(proteomics[c(downreg_genes_proteins_nearby_L1, "L1RE1"),levels(coldata$samples)], cluster_rows = T, cluster_cols = F, border_color = NA, annotation_row = all_downreg[downreg_genes_proteins_nearby_L1,"type", drop=F], annotation_col = coldata[,"condition", drop=F], scale = "row")

pheatmap(proteomics[tmp$tree_row$labels[tmp$tree_row$order],levels(coldata$samples)], cluster_rows = F, cluster_cols = F, border_color = NA, annotation_row = all_downreg[downreg_genes_proteins_nearby_L1,"type", drop=F], annotation_col = coldata[,"condition", drop=F])
# dev.off()
#
```

## Proteomics: ORF1p, pluripotency markers

```{r}
# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/heatmap_L1ORF1p_proteomics.pdf", height = 3, width = 12)
pheatmap(proteomics["L1RE1",levels(coldata$samples)], cluster_rows = F, cluster_cols = F, border_color = NA, labels_row = "L1-ORF1p", annotation_col = coldata[,"condition", drop=F], scale = "row")
pheatmap(proteomics["L1RE1",levels(coldata$samples)], cluster_rows = F, cluster_cols = F, border_color = NA, labels_row = "L1-ORF1p", annotation_col = coldata[,"condition", drop=F])
dev.off()

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/heatmap_pluripotency_proteomics.pdf", height = 6, width = 12)
pheatmap(proteomics[c("POU5F1","NANOG","UTF1","SOX2","ZFP42"),levels(coldata$samples)], cluster_rows = F, cluster_cols = F, border_color = NA,  annotation_col = coldata[,"condition", drop=F], scale = "row", main = "Pluripotency markers (scaled: row)")
pheatmap(proteomics[c("POU5F1","NANOG","UTF1","SOX2","ZFP42"),levels(coldata$samples)], cluster_rows = F, cluster_cols = F, border_color = NA, annotation_col = coldata[,"condition", drop=F], main = "Pluripotency markers")
# dev.off()

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/heatmap_L1ORF1p_proteomics.pdf", height = 3, width = 12)
pheatmap(proteomics[c("CTNNA3", "L1RE1"),levels(coldata$samples)], cluster_rows = F, cluster_cols = F, border_color = NA, annotation_col = coldata[,"condition", drop=F])
pheatmap(proteomics["L1RE1",levels(coldata$samples)], cluster_rows = F, cluster_cols = F, border_color = NA, labels_row = "L1-ORF1p", annotation_col = coldata[,"condition", drop=F])
# dev.off()
# 
# save.image("/Volumes/MyPassport/CRISPRi_L1s/src/r_scripts/gene_uniq_DEA.Rdata")
load("/Volumes/MyPassport/CRISPRi_L1s/src/r_scripts/gene_uniq_DEA.Rdata")
```

Not all are intersecting a >6kbp L1Hs-L1PA4... so, how close is the nearest one? (Fig4b-c)
```{r}
downreg_gene_hiPS_crispri_intersect_table <- data.frame(gene_name = downreg_gene_hiPS_crispri_type$gene_name,
           gene_type = downreg_gene_hiPS_crispri_type$gene_type,
           intersect_L1HS_PA4 = ifelse(downreg_gene_hiPS_crispri_type$gene_name %in% downreg_gene_hiPS_crispri_intersect$V4, T, F))

non_intersecting_L1s <- downreg_gene_hiPS_crispri_intersect_table[which(!downreg_gene_hiPS_crispri_intersect_table$intersect_L1HS_PA4),"gene_name"]
non_intersecting_L1s <- gene_bed[non_intersecting_L1s, ]
# nrow(read.table("/Volumes/MyPassport/CRISPRi_L1s/results/tables/downreg_gene_hiPS48_g3_crispri_nointersecting_L1.bed", ))
# write.table(non_intersecting_L1s[, c("chr", "start", "end", "gene_name", "dot", "strand")], "/Volumes/MyPassport/CRISPRi_L1s/results/tables/downreg_gene_hiPS48_g3_crispri_nointersecting_L1.bed", quote = F, sep="\t", col.names = F, row.names = F)

downreg_gene_closes_L1 <- read.table("/Volumes/MyPassport/CRISPRi_L1s/results/tables/downreg_gene_hiPS48_g3_crispri_nointersecting_L1_closes_6kbp_L1HS_L1PA4.bed")
colnames(downreg_gene_closes_L1) <- c("chr", "start", "end", "gene_name", "dot", "strand", "TE_chr", "TE_start", "TE_end", "TE_id", "TE_dot", "TE_strand", "TE_info", "distance")

library(plyr)
downreg_gene_closes_L1 <- rbind.fill(downreg_gene_closes_L1, gene_bed[downreg_gene_hiPS_crispri_intersect_table[which(as.logical(downreg_gene_hiPS_crispri_intersect_table$intersect_L1)), "gene_name"],])
downreg_gene_closes_L1[which(is.na(downreg_gene_closes_L1$distance)), "distance"] <- 0

colnames(downreg_gene_hiPS_crispri_intersect) <- c("chr", "start", "end", "gene_name", "dot", "strand", "TE_chr", "TE_start", "TE_end", "TE_id", "dot2", "TE_strand", "TE_info", "overlap")

downreg_gene_closes_L1 <- merge(downreg_gene_closes_L1, downreg_gene_hiPS_crispri_intersect, by=c("chr", "start", "end", "gene_name", "strand"), all.x=T)
downreg_gene_closes_L1$TE_chr.x <- ifelse(is.na(downreg_gene_closes_L1$TE_chr.x), downreg_gene_closes_L1$TE_chr.y, downreg_gene_closes_L1$TE_chr.x)
downreg_gene_closes_L1$TE_start.x <- ifelse(is.na(downreg_gene_closes_L1$TE_start.x), downreg_gene_closes_L1$TE_start.y, downreg_gene_closes_L1$TE_start.x)
downreg_gene_closes_L1$TE_end.x <- ifelse(is.na(downreg_gene_closes_L1$TE_end.x), downreg_gene_closes_L1$TE_end.y, downreg_gene_closes_L1$TE_end.x)
downreg_gene_closes_L1$TE_id.x <- ifelse(is.na(downreg_gene_closes_L1$TE_id.x), downreg_gene_closes_L1$TE_id.y, downreg_gene_closes_L1$TE_id.x)
downreg_gene_closes_L1$TE_id.x <- ifelse(is.na(downreg_gene_closes_L1$TE_id.x), downreg_gene_closes_L1$TE_id.y, downreg_gene_closes_L1$TE_id.x)
downreg_gene_closes_L1$TE_strand.x <- ifelse(is.na(downreg_gene_closes_L1$TE_strand.x), downreg_gene_closes_L1$TE_strand.y, downreg_gene_closes_L1$TE_strand.x)
downreg_gene_closes_L1$TE_info.x <- ifelse(is.na(downreg_gene_closes_L1$TE_info.x), downreg_gene_closes_L1$TE_info.y, downreg_gene_closes_L1$TE_info.x)
colnames(downreg_gene_closes_L1) <- c("chr","start","end","gene_name","strand","dot","TE_chr","TE_start","TE_end","TE_id","TE_dot","TE_strand", "TE_info","distance","gene_id","dot.y","TE_chr.y","TE_start.y","TE_end.y","TE_id.y","dot2","TE_strand.y", "TE_info.y","overlap")

downreg_gene_closes_L1 <- downreg_gene_closes_L1[,c("chr","start","end","gene_name","strand","TE_chr","TE_start","TE_end","TE_id","TE_strand", "TE_info","distance","overlap")]
downreg_gene_closes_L1 <- merge(downreg_gene_closes_L1, gene_type[downreg_gene_closes_L1$gene_name,])

downreg_gene_closes_L1 <- downreg_gene_closes_L1[which(!duplicated(downreg_gene_closes_L1[order(downreg_gene_closes_L1$gene_name, downreg_gene_closes_L1$distance, downreg_gene_closes_L1$overlap), "gene_name"])),]

downreg_gene_closes_L1$gene_type <- ifelse(downreg_gene_closes_L1$gene_type == "protein_coding", downreg_gene_closes_L1$gene_type, "ncRNA")
downreg_gene_closes_L1$gene_type <- paste(downreg_gene_closes_L1$gene_type, ifelse(downreg_gene_closes_L1$distance == 0, "L1Overlap", "L1Nearby"), sep="_")
downreg_gene_closes_L1$gene_type <- factor(downreg_gene_closes_L1$gene_type, levels = c("protein_coding_L1Overlap", "protein_coding_L1Nearby", "ncRNA_L1Overlap", "ncRNA_L1Nearby"))

downreg_gene_closes_L1$distance_50kb <- ifelse(downreg_gene_closes_L1$distance > 50000, ">50kb", "=<50kb")
downreg_gene_closes_L1$overlap_type <- ifelse(grepl("Overlap", downreg_gene_closes_L1$gene_type), "Overlap", "Nearby")

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/hist_downreg_genes_distance_nearest_6kbp_L1HS_L1PA4.pdf", height = 3, width = 6)
ggplot(downreg_gene_closes_L1, aes(x=distance, fill=gene_type)) + geom_histogram() + 
  theme_bw() + 
  theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank()) + 
  labs(x="Distance to nearest >6kbp L1HS-L1PA4", y="Count") + lims(y=c(0,60))
# dev.off()

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/hist_downreg_genes_distance_nearest_6kbp_L1HS_L1PA4_50kb.pdf", height = 4, width = 4)
ggplot(downreg_gene_closes_L1, aes(x=distance_50kb, fill=overlap_type)) + geom_bar(stat="count") + 
  theme_bw() + 
  theme(panel.grid.minor = element_blank(), panel.grid.major = element_blank()) + 
  labs(x="Distance to nearest >6kbp L1HS-L1PA4", y="Count", fill="") + ggtitle("Downregulated genes")
# dev.off()

genes_down_nearby_L1 <- downreg_gene_closes_L1[which(downreg_gene_closes_L1$distance_50kb == "=<50kb"), ]
genes_down_nearby_L1$gene_type <- as.character(genes_down_nearby_L1$gene_type)
genes_down_nearby_L1$gene_type <- ifelse(startsWith(genes_down_nearby_L1$gene_type, "protein_coding"), "protein_coding", "ncRNA")
genes_down_nearby_L1 <- genes_down_nearby_L1[order(genes_down_nearby_L1$gene_type),]
rownames(genes_down_nearby_L1) <- genes_down_nearby_L1$gene_name

# pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/heatmap_downreg_genes_intersect_l1hs_l1pa4_hiPS48_g3_50kb.pdf", height = 15)
tmp_heatmap <- pheatmap(gene_count_norm[genes_down_nearby_L1$gene_name, c(samplesheet_comparisons$hiPS48_crispri_g3$samples,
                                                        samplesheet_comparisons$hiPS6_crispri_g3$samples,
                                                        samplesheet_comparisons$hiPS6_crispri_g1$samples)], scale = "row", cluster_cols = F, cluster_rows = F, border_color = NA, annotation_row = genes_down_nearby_L1[, "gene_type", drop=F])

pheatmap(gene_count_norm[genes_down_nearby_L1$gene_name, c(samplesheet_comparisons$hiPS48_crispri_g3$samples)], scale = "row", cluster_cols = F, cluster_rows = F, border_color = NA, main = "hiPS48_crispri_g3", breaks = seq(-2,2, length = 100), annotation_row = genes_down_nearby_L1[, "gene_type", drop=F])
pheatmap(gene_count_norm[genes_down_nearby_L1$gene_name, c(samplesheet_comparisons$hiPS6_crispri_g3$samples)], scale = "row", cluster_cols = F, cluster_rows = F, border_color = NA, main = "hiPS6_crispri_g3", breaks = seq(-2,2, length = 100), annotation_row = genes_down_nearby_L1[, "gene_type", drop=F])
pheatmap(gene_count_norm[genes_down_nearby_L1$gene_name, c(samplesheet_comparisons$hiPS6_crispri_g1$samples)], scale = "row", cluster_cols = F, cluster_rows = F, border_color = NA, main = "hiPS6_crispri_g1", breaks = seq(-2,2, length = 100), annotation_row = genes_down_nearby_L1[, "gene_type", drop=F])
dev.off()
```

