---
title: "CRISPR L1s differential expression analysis of TEs"
output: html_notebook
---

Here we perform differential expression analysis (DEA) of all TEs using a unique mapping approach for our CRISPR experiments targeting young L1 elements.

The experimental design is CRISPR inhibition on iPSC where L1s are usually highly expressed, and CRISPR activation on neural progenitor cells.
The two experiments were sequenced in two sequencing runs each (three sequencing runs total). 

Here we load the metadata of the samples (and split them by experiment and guide), and the TE count matrices (`te_counts` with all counts).
```{r}
library(data.table)
library(ggplot2)
library(ggpubr)
library(stringr)
library(DESeq2)

samplesheet <- fread("/Volumes/MyPassport/CRISPRi_L1s/samplesheet_L1crispr.tab", data.table = F, header = T)
samplesheet <- samplesheet[which(!samplesheet$samples %in% c("AA11_Sai2_lv3775_B_CRISPRa_S11", "AA_Sai2_L1_CRISPRa_LV3776_c_S29")),]
samplesheet_list <- split(samplesheet, f = list(samplesheet$cell, samplesheet$experiment, samplesheet$guide, samplesheet$seqnum))
samplesheet_list <- samplesheet_list[sapply(samplesheet_list,nrow)>0] 

samplesheet_comparisons <- list(
  "hiPS6_crispri_g1" = rbind(samplesheet_list$hiPS6.crispri.g1.CTG_JGJSeq156_159_160truseq, 
                             samplesheet_list$hiPS6.crispri.LacZ.CTG_JGJSeq156_159_160truseq),
  "hiPS6_crispri_g2" = rbind(samplesheet_list$hiPS6.crispri.g2.CTG_JGJSeq205_207_208_214_truseq_2023_021,
                             samplesheet_list$hiPS6.crispri.LacZ.CTG_JGJSeq205_207_208_214_truseq_2023_021),
  "hiPS6_crispri_g3" = rbind(samplesheet_list$hiPS6.crispri.g3.CTG_JGJSeq205_207_208_214_truseq_2023_021,
                             samplesheet_list$hiPS6.crispri.LacZ.CTG_JGJSeq205_207_208_214_truseq_2023_021),
  "hiPS48_crispri_g3" = rbind(samplesheet_list$hiPS48.crispri.g3.testdata_ctg_2023_50,
                              samplesheet_list$hiPS48.crispri.LacZ.testdata_ctg_2023_50),
  "NES_crispra_g1" = rbind(samplesheet_list$NES.crispra.g1.CTG_JGJSeq156_159_160truseq,
                           samplesheet_list$NES.crispra.NTC.CTG_JGJSeq156_159_160truseq),
  "NES_crispra_g2" = rbind(samplesheet_list$NES.crispra.g2.CTG_JGJSeq233_237_250_257_2023_145,
                           samplesheet_list$NES.crispra.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "NES_crispra_g3" = rbind(samplesheet_list$NES.crispra.g3.CTG_JGJSeq233_237_250_257_2023_145,
                           samplesheet_list$NES.crispra.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "hiPS48_org_crispri_g3" = rbind(samplesheet_list$hiPS48_org.crispri.g3.CTGseq278_280_281_282_285_288_289_290_291_2023_282NovaseqX,
                                  samplesheet_list$hiPS48_org.crispri.LacZ.CTGseq278_280_281_282_285_288_289_290_291_2023_282NovaseqX),
  "hiPS6_org_crispri_g3" = rbind(samplesheet_list$hiPS6_org.crispri.g3.CTG_JGJSeq233_237_250_257_2023_145,
                                  samplesheet_list$hiPS6_org.crispri.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "hiPS6_org_crispra_g3" = rbind(samplesheet_list$hiPS6_org.crispra.g3.CTG_JGJSeq233_237_250_257_2023_145,
                                  samplesheet_list$hiPS6_org.crispra.LacZ.CTG_JGJSeq233_237_250_257_2023_145),
  "hiPS6_org_crispri_g1" = rbind(samplesheet_list$hiPS6_org.crispri.g1.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispri.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispri_g2" = rbind(samplesheet_list$hiPS6_org.crispri.g2.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispri.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispri_g3_batch2" = rbind(samplesheet_list$hiPS6_org.crispri.g3.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispri.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispra_g1" = rbind(samplesheet_list$hiPS6_org.crispra.g1.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispra.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispra_g2" = rbind(samplesheet_list$hiPS6_org.crispra.g2.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispra.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq),
  "hiPS6_org_crispra_g3_batch2" = rbind(samplesheet_list$hiPS6_org.crispra.g3.CTG_2024_017_Seq296_297_305_315_bulkRNASeq,
                                 samplesheet_list$hiPS6_org.crispra.LacZ.CTG_2024_017_Seq296_297_305_315_bulkRNASeq)
)

path <- "/Volumes/MyPassport/CRISPRi_L1s/TEcounts/unique_strand/"
samples <- samplesheet$samples

sizefactors <- fread("/Volumes/MyPassport/CRISPRi_L1s/results/tables/gene_sizeFactors.tab", data.table = F)
colnames(sizefactors) <- c("sample", "sizeFactor")
rownames(sizefactors) <- sizefactors$sample
for(i in 1:length(samples)){
  sample <- samples[i]
  if(i == 1){
    te_counts <- fread(paste(path, sample, "_TE_count_matrix_0.csv", sep=""), data.table = F)    
    if(grepl(pattern = "forward", colnames(te_counts)[ncol(te_counts)-1]) & grepl(pattern = "reverse", colnames(te_counts)[ncol(te_counts)])){
      colnames(te_counts)[c((ncol(te_counts)-1), ncol(te_counts))] <- c(paste(sample, "_forward", sep=""), paste(sample, "_reverse", sep=""))  
    }
    te_counts[,c((ncol(te_counts)-1), ncol(te_counts))] <- te_counts[,c((ncol(te_counts)-1), ncol(te_counts))] / sizefactors[sample,"sizeFactor"]
    rownames(te_counts) <- te_counts$Geneid
    row_order <- rownames(te_counts)
  }else{
    tmp <- fread(paste(path, sample, "_TE_count_matrix_0.csv", sep=""), data.table = F)
    if(grepl(pattern = "forward", colnames(tmp)[ncol(tmp)-1]) & grepl(pattern = "reverse", colnames(tmp)[ncol(tmp)])){
      colnames(tmp)[c((ncol(tmp)-1), ncol(tmp))] <- c(paste(sample, "_forward", sep=""), paste(sample, "_reverse", sep=""))  
    }
    tmp[,c((ncol(tmp)-1), ncol(tmp))] <- tmp[,c((ncol(tmp)-1), ncol(tmp))] / sizefactors[sample,"sizeFactor"]
    rownames(tmp) <- tmp$Geneid
    te_counts <- cbind(te_counts[row_order,], tmp[row_order, c(paste(sample, "_forward", sep=""), paste(sample, "_reverse", sep="")), drop=F])
  }
}

```

```{r}
te_counts$TE_subfam <- sapply(str_split(te_counts$Geneid, "_dup"), `[[`, 1)

fwd_strand_te_counts <- te_counts[which(te_counts$Strand == "+"), ]
colnames(fwd_strand_te_counts) <- str_replace_all(colnames(fwd_strand_te_counts), pattern = "forward", replacement = "sense")
colnames(fwd_strand_te_counts) <- str_replace_all(colnames(fwd_strand_te_counts), pattern = "reverse", replacement = "antisense")

rev_strand_te_counts <- te_counts[which(te_counts$Strand == "-"), ]
colnames(rev_strand_te_counts) <- str_replace_all(colnames(rev_strand_te_counts), pattern = "reverse", replacement = "sense")
colnames(rev_strand_te_counts) <- str_replace_all(colnames(rev_strand_te_counts), pattern = "forward", replacement = "antisense")

sense_te_counts <- rbind(rev_strand_te_counts[,colnames(fwd_strand_te_counts)], fwd_strand_te_counts[,colnames(fwd_strand_te_counts)])

sense_te_counts_aggr_TEsubfam <- aggregate(sense_te_counts[, 7:(ncol(te_counts)-1)], by=list(sense_te_counts$TE_subfam), FUN=sum)
L1PA_sense_te_counts_aggr_TEsubfam <- sense_te_counts_aggr_TEsubfam[which(sense_te_counts_aggr_TEsubfam$Group.1 %in% c("L1HS", "L1PA2", "L1PA3", "L1PA4", "L1PA5", "L1PA6", "L1PA7")),]

L1PA_sense_te_counts_aggr_TEsubfam_melt <- reshape2::melt(L1PA_sense_te_counts_aggr_TEsubfam, by=list(L1PA_sense_te_counts_aggr_TEsubfam$Group.1))
L1PA_sense_te_counts_aggr_TEsubfam_melt$direction <- ifelse(grepl("antisense", L1PA_sense_te_counts_aggr_TEsubfam_melt$variable), "Antisense", "Sense")

data_summary <- function(data, varname, groupnames){
  require(plyr)
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE),
      sd = sd(x[[col]], na.rm=TRUE))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func,
                  varname)
  data_sum <- rename(data_sum, c("mean" = varname))
 return(data_sum)
}

sense_boxplots <- list()
for(i in names(samplesheet_comparisons)){
  comparison <- samplesheet_comparisons[[i]]
  te_counts_sample_tmp <- L1PA_sense_te_counts_aggr_TEsubfam_melt[which(L1PA_sense_te_counts_aggr_TEsubfam_melt$variable %in% c(paste(comparison$samples, "_sense", sep=""), paste(comparison$samples, "_antisense", sep=""))),]
  te_counts_sample_tmp$samples <- str_remove_all(te_counts_sample_tmp$variable, "_sense")
  te_counts_sample_tmp$samples <- str_remove_all(te_counts_sample_tmp$sample, "_antisense")
  te_counts_sample_tmp <- merge(te_counts_sample_tmp, comparison, by="samples")
  sense_boxplots[[i]] <- list()
  for(L1 in c("L1HS", "L1PA2", "L1PA3", "L1PA4", "L1PA5", "L1PA6", "L1PA7")){
    te_counts_sample_L1_tmp <- te_counts_sample_tmp[which(te_counts_sample_tmp$Group.1 == L1),]
    max_lim <- max(te_counts_sample_L1_tmp$value) + mean(te_counts_sample_L1_tmp$value)/2
    control_df <- te_counts_sample_L1_tmp[which(te_counts_sample_L1_tmp$condition == "Control"), ]
    effect_df <- te_counts_sample_L1_tmp[which(te_counts_sample_L1_tmp$condition != "Control"), ]
    control_df_stats <- data_summary(control_df, varname="value", groupnames=c("direction"))
    effect_df_stats <- data_summary(effect_df, varname="value", groupnames=c("direction"))
    
    sense_boxplots[[i]][[L1]][["Control"]] <- 
      ggplot(data = control_df, aes(x=direction, y=value)) +
      geom_bar(data = control_df_stats, aes(x=direction, y=value, fill=direction), stat="identity", alpha=0.7) + 
      geom_errorbar(data = control_df_stats, aes(ymin=value-sd, ymax=value+sd), color="black", alpha=0.8, width=.2,position=position_dodge(0.05)) +
      geom_line(data = control_df, aes(group=samples), alpha = 0.3) +
      geom_point(color="black", alpha=0.7) +
      stat_compare_means(comparisons = list(c("Antisense", "Sense")), label = "p.format", method = "t.test", paired=T, tip.length = 0) +
      scale_fill_manual(values = c("Antisense" = "tomato", "Sense" = "darkturquoise")) + facet_wrap(.~condition) + lims(y=c(0, max_lim)) +
      theme_bw() + labs(x="", y="Accumulated expression", fill="") + ggtitle(paste(i, ":", L1)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), plot.title = element_text(size = 6))
    
    
    sense_boxplots[[i]][[L1]][[unique(effect_df$condition)]] <- 
      ggplot(data = effect_df, aes(x=direction, y=value)) +
      geom_bar(data = effect_df_stats, aes(x=direction, y=value, fill=direction), stat="identity", alpha=0.7) + 
      geom_errorbar(data = effect_df_stats, aes(ymin=value-sd, ymax=value+sd), color="black", alpha=0.8, width=.2,position=position_dodge(0.05)) +
      geom_line(data = effect_df, aes(group=samples), alpha = 0.3) +
      geom_point(color="black", alpha=0.7) +
      stat_compare_means(comparisons = list(c("Antisense", "Sense")), label = "p.format", method = "t.test", paired=T, tip.length = 0) +
      scale_fill_manual(values = c("Antisense" = "tomato", "Sense" = "darkturquoise")) + facet_wrap(.~condition) + lims(y=c(0, max_lim)) +
      theme_bw() + labs(x="", y="Accumulated expression", fill="") + ggtitle(paste(i, ":", L1)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), plot.title = element_text(size = 6))
  }
}

comparison_wt <- rbind(samplesheet_list$hiPS48_WT.WT_iPSC.WT.CTG_Seq107_111_112, 
                       samplesheet_list$hiPS6_WT.WT_iPSC.WT.CTG_Seq107_111_112)

te_counts_sample_tmp <- L1PA_sense_te_counts_aggr_TEsubfam_melt[which(L1PA_sense_te_counts_aggr_TEsubfam_melt$variable %in% c(paste(comparison_wt$samples, "_sense", sep=""), paste(comparison_wt$samples, "_antisense", sep=""))),]
te_counts_sample_tmp$samples <- str_remove_all(te_counts_sample_tmp$variable, "_sense")
te_counts_sample_tmp$samples <- str_remove_all(te_counts_sample_tmp$sample, "_antisense")
te_counts_sample_tmp <- merge(te_counts_sample_tmp, comparison_wt, by="samples")
sense_boxplots[["hiPS_WT"]] <- list()
for(L1 in c("L1HS", "L1PA2", "L1PA3", "L1PA4", "L1PA5", "L1PA6", "L1PA7")){
  te_counts_sample_L1_tmp <- te_counts_sample_tmp[which(te_counts_sample_tmp$Group.1 == L1),]
  max_lim <- max(te_counts_sample_L1_tmp$value) + mean(te_counts_sample_L1_tmp$value)/2
  hips6_df <- te_counts_sample_L1_tmp[which(te_counts_sample_L1_tmp$cell == "hiPS6_WT"), ]
  hips48_df <- te_counts_sample_L1_tmp[which(te_counts_sample_L1_tmp$cell == "hiPS48_WT"), ]
  hips6_df_stats <- data_summary(hips6_df, varname="value", groupnames=c("direction"))
  hips48_df_stats <- data_summary(hips48_df, varname="value", groupnames=c("direction"))
  
  sense_boxplots[["hiPS_WT"]][[L1]][["hiPS6_WT"]] <- 
    ggplot(data = hips6_df, aes(x=direction, y=value)) +
    geom_bar(data = hips6_df_stats, aes(x=direction, y=value, fill=direction), stat="identity", alpha=0.7) + 
    geom_errorbar(data = hips6_df_stats, aes(ymin=value-sd, ymax=value+sd), color="black", alpha=0.8, width=.2,position=position_dodge(0.05)) +
    geom_line(data = hips6_df, aes(group=samples), alpha = 0.3) +
    geom_point(color="black", alpha=0.7) +
    stat_compare_means(comparisons = list(c("Antisense", "Sense")), label = "p.format", method = "t.test", paired=T, tip.length = 0) +
    scale_fill_manual(values = c("Antisense" = "tomato", "Sense" = "darkturquoise")) + facet_wrap(.~condition) + lims(y=c(0, max_lim)) +
    theme_bw() + labs(x="", y="Accumulated expression", fill="") + ggtitle(paste("hiPS6_WT:", L1)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), plot.title = element_text(size = 6))
  
  sense_boxplots[["hiPS_WT"]][[L1]][["hiPS48_WT"]] <- 
    ggplot(data = hips48_df, aes(x=direction, y=value)) +
    geom_bar(data = hips48_df_stats, aes(x=direction, y=value, fill=direction), stat="identity", alpha=0.7) + 
    geom_errorbar(data = hips48_df_stats, aes(ymin=value-sd, ymax=value+sd), color="black", alpha=0.8, width=.2,position=position_dodge(0.05)) +
    geom_line(data = hips48_df, aes(group=samples), alpha = 0.3) +
    geom_point(color="black", alpha=0.7) +
    stat_compare_means(comparisons = list(c("Antisense", "Sense")), label = "p.format", method = "t.test", paired=T, tip.length = 0) +
    scale_fill_manual(values = c("Antisense" = "tomato", "Sense" = "darkturquoise")) + facet_wrap(.~condition) + lims(y=c(0, max_lim)) +
    theme_bw() + labs(x="", y="Accumulated expression", fill="") + ggtitle(paste("hiPS48_WT:", L1)) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), plot.title = element_text(size = 6))
}

```


```{r}
sense_boxplots_arranged <- list()
for(i in names(sense_boxplots)){
  for(L1 in names(sense_boxplots[[i]])){
    print(i)
    print(L1)
    sense_boxplots_arranged[[i]][[L1]] <- ggarrange(plotlist = sense_boxplots[[i]][[L1]], ncol=2, nrow=1, common.legend = T)  
  }
}

order <- c(
  "hiPS_WT",
  "hiPS6_crispri_g1",
  "hiPS6_crispri_g2",
  "hiPS6_crispri_g3",
  "hiPS48_crispri_g3",
  "NES_crispra_g1",
  "NES_crispra_g2",
  "NES_crispra_g3",
  "hiPS6_org_crispri_g1",
  "hiPS6_org_crispri_g2",
  "hiPS48_org_crispri_g3",
  "hiPS6_org_crispri_g3_batch2",
  "hiPS6_org_crispri_g3",
  "hiPS6_org_crispra_g1",
  "hiPS6_org_crispra_g2",
  "hiPS6_org_crispra_g3",
  "hiPS6_org_crispra_g3_batch2")

pdf("/Volumes/MyPassport/CRISPRi_L1s/results/plots/barplots_direction.pdf", height = 4, width = 6)
sense_boxplots_arranged[order]
dev.off()

save.image("/Volumes/MyPassport/CRISPRi_L1s/src/r_scripts/TE_uniq_direction.Rdata")
```
```{r}

```

